src/core/controllers/map-controller.js

```diff
@@ -0,0 +1,286 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Controller from './controller';
+import WebMercatorViewport from '../viewports/web-mercator-viewport';
+import {normalizeViewportProps} from 'viewport-mercator-project';
+
+import {clamp} from 'math.gl';
+import assert from 'assert';
+
+// At the moment we enforce mapbox limits by default
+import MAPBOX_LIMITS from './mapbox-limits';
+
+// GENERIC MAP LIMITS
+export const MAP_LIMITS = {
+  // latitude which corresponds to a projected square (y = 0), typical limit for maps
+  maxLatitude: 85.05113,
+  minLatitude: -85.05113,
+  minPitch: 0,
+  maxPitch: 89,
+  minZoom: 0,
+  maxZoom: 40
+};
+
+// EVENT HANDLING PARAMETERS
+const PITCH_MOUSE_THRESHOLD = 5;
+const PITCH_ACCEL = 1.2;
+
+export function mod(value, divisor) {
+  const modulus = value % divisor;
+  return modulus < 0 ? divisor + modulus : modulus;
+}
+
+// Helper functions
+function normalizeDegrees(degrees) {
+  return degrees < -180 || degrees > 180 ? mod(degrees + 180, 360) - 180 : degrees;
+}
+
+// EVENT REDUCERS
+
+// Pan
+// @param {[Number, Number]} pos - position on screen where the pointer is
+// @param {[Number, Number], optional} startPos - where the pointer grabbed at
+//   the start of the operation. Must be supplied of `panStart()` was not called
+function pan(viewState, {pos, startPos, startViewState}) {
+  const startViewport = new WebMercatorViewport(startViewState.props);
```

Comment:
Need to check if `startViewState` is empty, in case a child component captures the `panstart` event.